File: java-app/src/main/java/com/terpomo/wavy/flow/Project.java
package com.terpomo.wavy.flow;

import com.terpomo.wavy.core.IWavyModel;
import com.terpomo.wavy.marshal.IMarshallable;
import com.terpomo.wavy.marshal.MarshalAttr;
import com.terpomo.wavy.marshal.MarshallingKeys;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;

public class Project implements IWavyModel, IMarshallable {

	private static final Logger LOGGER = Logger.getLogger(Project.class.getName());
	private String name;
	private List<IPipe> pipes;
	private Map<String, IPipe> nameToPipeMap;
	
	public Project() {
		this.name = "";
		this.pipes = new ArrayList<IPipe>();
		this.nameToPipeMap = new HashMap<>();
	}

	@MarshalAttr(attrName = MarshallingKeys.KEY_NAME)
	public String getName() {
		return name;
	}

	@MarshalAttr(attrName = MarshallingKeys.KEY_NAME)
	public void setName(String name) {
		this.name = name;
	}

	public List<IPipe> getPipes() {
		return pipes;
	}

	public synchronized void addPipe(IPipe p) {
		if (!this.getPipes().contains(p)) {
			ArrayList<IPipe> newPipes = new ArrayList<IPipe>(this.getPipes());
			newPipes.add(p);
			this.setPipes(newPipes);
		}
	}

	public synchronized void removePipe(IPipe pipe) {
		ArrayList<IPipe> newPipes = new ArrayList<IPipe>(this.getPipes());
		newPipes.remove(pipe);
		this.setPipes(newPipes);
	}

	private void setPipes(List<IPipe> pipes) {
		this.pipes = pipes;
		this.rebuildNameToPipeMap(this.pipes);
	}

	public IPipe getPipeByName(String name) {
		return this.nameToPipeMap.get(name);
	}

	private void rebuildNameToPipeMap(List<IPipe> pipes) {
		Map<String, IPipe> map = new HashMap<>();
		for (IPipe pipe : pipes) {
			if (map.containsKey(pipe.getName()))
				LOGGER.warning(String.format("Pipe name conflict within project %s. Pipe name %s", this.getName(), pipe.getName()));
			map.put(pipe.getName(), pipe);
		}
		this.nameToPipeMap = map;
	}
}

File: wpf-app/Flow/Project.cs
using Wavy.Pipes;

namespace Wavy.Flow
{
    public class Project
    {
        public delegate void PipesModifiedEventHandler(object sender, PipesEventArgs e);
        public event PipesModifiedEventHandler? OnPipeAdded;
        public event PipesModifiedEventHandler? OnPipeRemoved;
        private HashSet<Pipe> _Pipes;
        public Project()
        {
            this.Name = String.Format("Project {0}", Random.Shared.Next());
            this._Pipes = new HashSet<Pipe>();
        }
        public string Name { get; set; }

        public void AddPipe(Pipe pipe)
        {
            if (pipe != null && !this._Pipes.Contains(pipe))
            {
                this._Pipes.Add(pipe);
                this.OnPipeAdded?.Invoke(this, new PipesEventArgs(pipe));
            }
        }

        public void RemovePipe(Pipe pipe) {
            if (this._Pipes.Contains(pipe))
            {
                this._Pipes.Remove(pipe);
                this.OnPipeRemoved?.Invoke(this, new PipesEventArgs(pipe));
            }
        }
    }

    public class PipesEventArgs : EventArgs
    {
        public Pipe Pipe { get; private set; }

        public PipesEventArgs(Pipe pipe) {
            this.Pipe = pipe;
        }
    }
}
